staging/src/k8s.io/apiserver/pkg/admission/metrics.go

```diff
@@ -0,0 +1,151 @@
+		"Admission sub-step %s, broken out for each operation and API resource and step type (validate or admit).")
+
+	// Built-in admission controller metrics. Each admission controller is identified by name.
+	controller := newMetricSet("controller",
+		[]string{"name", "type", "operation", "group", "version", "resource", "subresource", "rejected"},
+		"Admission controller %s, identified by name and broken out for each operation and API resource and type (validate or admit).")
+
+	// Admission webhook metrics. Each webhook is identified by name.
+	webhook := newMetricSet("webhook",
+		[]string{"name", "type", "operation", "group", "version", "resource", "subresource", "rejected"},
+		"Admission webhook %s, identified by name and broken out for each operation and API resource and type (validate or admit).")
+
+	step.mustRegister()
+	controller.mustRegister()
+	webhook.mustRegister()
+	return &AdmissionMetrics{step: step, controller: controller, webhook: webhook}
+}
+
+func (m *AdmissionMetrics) reset() {
+	m.step.reset()
+	m.controller.reset()
+	m.webhook.reset()
+}
+
+// ObserveAdmissionStep records admission related metrics for a admission step, identified by step type.
+func (m *AdmissionMetrics) ObserveAdmissionStep(elapsed time.Duration, rejected bool, attr Attributes, stepType string) {
+	gvr := attr.GetResource()
+	m.step.observe(elapsed, stepType, string(attr.GetOperation()), gvr.Group, gvr.Version, gvr.Resource, attr.GetSubresource(), strconv.FormatBool(rejected))
+}
+
+// ObserveAdmissionController records admission related metrics for a built-in admission controller, identified by it's plugin handler name.
+func (m *AdmissionMetrics) ObserveAdmissionController(elapsed time.Duration, rejected bool, handler NamedHandler, attr Attributes, stepType string) {
+	gvr := attr.GetResource()
+	m.controller.observe(elapsed, handler.Name(), stepType, string(attr.GetOperation()), gvr.Group, gvr.Version, gvr.Resource, attr.GetSubresource(), strconv.FormatBool(rejected))
+}
+
+// ObserveWebhook records admission related metrics for a admission webhook.
+func (m *AdmissionMetrics) ObserveWebhook(elapsed time.Duration, rejected bool, hook *v1alpha1.Webhook, attr Attributes) {
+	t := "admit" // TODO: pass in type (validate|admit) once mutating webhook functionality has been implemented
+	gvr := attr.GetResource()
+	m.webhook.observe(elapsed, hook.Name, t, string(attr.GetOperation()), gvr.Group, gvr.Version, gvr.Resource, attr.GetSubresource(), strconv.FormatBool(rejected))
+}
+
+type metricSet struct {
+	latencies        *prometheus.HistogramVec
+	latenciesSummary *prometheus.SummaryVec
+}
+
+func newMetricSet(name string, labels []string, helpTemplate string) *metricSet {
+	return &metricSet{
+		latencies: prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      fmt.Sprintf("%s_latencies_seconds", name),
```

Comment:
I still think the metric name is a wrong place to make the distinction, essentially this is the same metric just from different places. `Subsystem` is a much more suitable name for this case

To me, `adminssions_controller_admissions_latencies_seconds`  sounds better than `adminssions_controller_latencies_seconds`, because it's clear what's the metric is about. If a new metric appears, e.g. `adminssions_controller_doing_stuff_latencies_seconds`, it won't be consistent. Additionally, when exporting to other systems, e.g. Stackdriver Monitoring, making metric name itself self-contained makes even more sense

Reply:
@fabxc @piosz WDYT? Let's have some quorum here

Reply:
Also @fgrzadkowski

We should totally create K8s guidelines for metrics

Reply:
`adminssions_controller_admissions_latencies_seconds` is what we had originally proposed in the design before adjustments based on feedback, so if we can get agreement on that as an instrumentation naming convention going forward, I'm more than happy to change it back.

Reply:
@crassirostris I don't understand the name vs. subsystem suggestion. What are you proposing?

Reply:
Effectively that would be the same exposed name, but in the code that would look like this:

```go
prometheus.NewHistogramVec(
	prometheus.HistogramOpts{
		Namespace: namespace,
		Subsystem: fmt.Sprintf("admission_%s", name),
		Name:      "admission_latencies_seconds",
		Help:      fmt.Sprintf(helpTemplate, "latency histogram"),
		Buckets:   latencyBuckets,
	},
	labels,
),
```

So it looks like the same metric, but for different components: admission webhook, admission controller, admission step.

Another way I find cleaner than the current solution would be to have a metric named `apiserver_admission_latencies_seconds` that has a label `component` or `source`

Reply:
I've updated the metric names to include _admission_.

We need to keep the existing conmponents separate to adhere to the "[metric names] should represent the same logical thing-being-measured across all label dimensions" prometheus guideline. In this case, steps, controllers and webhooks are all logically distinct-- components are compiled in code while webhooks are called remotely... If combined, a unfiltered query against the combined metric is at risk of being more misleading than helpful, also suggesting they should be kept logically separate.

Subsystem does not strike me as a good field for component names. `admission` is the logical subsystem here, as far as I can tell the hierarchy we use for metrics is something like service/subsystem/component/metric.  Which suggests keeping subsystem as is. It's also an implementation detail and doesn't impact the exposed metric name.

Reply:
Thanks!

> If combined, a unfiltered query against the combined metric is at risk of being more misleading than helpful

Sounds reasonable